adafruit_display_shapes/polygon.py

```diff
@@ -33,121 +33,162 @@
+        x_offset = min(x_s)
+        y_offset = min(y_s)
 
         # Find the largest and smallest X values to figure out width for bitmap
-        width = max(xs) - min(xs) + 1
-        height = max(ys) - min(ys) + 1
+        width = max(x_s) - min(x_s) + 1
+        height = max(y_s) - min(y_s) + 1
 
-        self._palette = displayio.Palette(3)
+        self._palette = displayio.Palette(colors + 1)
         self._palette.make_transparent(0)
-        self._bitmap = displayio.Bitmap(width, height, 3)
+        self._bitmap = displayio.Bitmap(width, height, colors + 1)
+
+        shifted = [(x - x_offset, y - y_offset) for (x, y) in points]
 
         if outline is not None:
-            # print("outline")
             self.outline = outline
-            for index, _ in enumerate(points):
-                point_a = points[index]
-                if index == len(points) - 1:
-                    point_b = points[0]
-                else:
-                    point_b = points[index + 1]
-                self._line(
-                    point_a[0] - x_offset,
-                    point_a[1] - y_offset,
-                    point_b[0] - x_offset,
-                    point_b[1] - y_offset,
-                    1,
-                )
+            self.draw(self._bitmap, shifted, self._OUTLINE, close)
 
         super().__init__(
             self._bitmap, pixel_shader=self._palette, x=x_offset, y=y_offset
         )
 
-    # pylint: disable=invalid-name, too-many-locals, too-many-branches
+    @staticmethod
+    def draw(
+        bitmap: displayio.Bitmap,
+        points: List[Tuple[int, int]],
+        color_id: int,
+        close: Optional[bool] = True,
+    ) -> None:
+        """Draw a polygon conecting points on provided bitmap with provided color_id
+
+        :param displayio.Bitmap bitmap: bitmap to draw on
+        :param list points: A list of (x, y) tuples of the points
+        :param int color_id: Color to draw with
+        :param bool close: (Optional) Wether to connect first and last point. (True)
+        """
+
+        if close:
+            points.append(points[0])
+
+        for index in range(len(points) - 1):
+            Polygon._line_on(bitmap, points[index], points[index + 1], color_id)
+
+    # pylint: disable=too-many-arguments
     def _line(
         self,
-        x0: int,
-        y0: int,
-        x1: int,
-        y1: int,
+        x_0: int,
+        y_0: int,
+        x_1: int,
+        y_1: int,
+        color: int,
+    ) -> None:
+        self._line_on(self._bitmap, (x_0, y_0), (x_1, y_1), color)
+
+    # pylint: enable=too-many-arguments
+
+    @staticmethod
+    def _safe_draw(
+        bitmap: displayio.Bitmap,
+        point: Tuple[int, int],
+        color: int,
+    ) -> None:
+        (x, y) = point
+        if 0 <= x < bitmap.width and 0 <= y < bitmap.height:
+            bitmap[x, y] = color
+
+    # pylint: disable=too-many-branches, too-many-locals
+    @staticmethod
+    def _line_on(
```

Comment:
I think it may be possible to use this built-in bitmaptools function from the core rather than having the implementation in python here. https://docs.circuitpython.org/en/latest/shared-bindings/bitmaptools/index.html#bitmaptools.draw_line

Perhaps ultimately we could make a `draw_polygon()` function inside bitmaptools as well which could then be used generally from anywhere, including this library.

Reply:
It's great to see that built-in function, I didn't knew it exists! Feel free to approve/merge this MR already, I'll work on using built-in function either way - one doubt I have is handling out-of-bonds coordinates by it, I'll need to check how it behaves.

As for adding `draw_polygon()` to bitmaptools I'm totally up for it, sounds like a nice starting point for contribting to circuitpython core (-:

Reply:
I just checked, `bitmaptools.draw_line()` does not like lines going out of bitmap - I think that it could be a valid scenario, so I may end up trying to change that behaviour in `bitmaptools` before using it here.